packages/core/src/storage/extract.ts

```diff
@@ -17,62 +22,98 @@ export function extractStorageLayout(
   contractDef: ContractDefinition,
   decodeSrc: SrcDecoder,
   deref: ASTDereferencer,
+  storageLayout?: StorageLayout | undefined,
 ): StorageLayout {
-  const layout: StorageLayout = { storage: [], types: {}, layoutVersion: currentLayoutVersion };
+  const layout: StorageLayout = { storage: [], types: {}, layoutVersion: currentLayoutVersion, flat: false };
+  if (storageLayout !== undefined) {
+    layout.types = storageLayout.types;
+    if (layout.types) {
+      for (const type of Object.entries(layout.types)) {
```

Comment:
maybe it would be clearer to for

```
for (const [ key, type ] of Object.entries(layout.type)) { 
```

You could even remove line 30 by doing
```
for (const [ key, type ] of Object.entries(layout.type || {})) { 
```